feat: crud of network members of a person

jnizet opened the pull request.

cexbrayat commented on frontend/src/app/person-network-members/person-network-members.component.spec.ts:
```diff
@@ -0,0 +1,257 @@
+
+  it('should delete a member and reload list after confirmation', () => {
+    const confirmService: ConfirmService = TestBed.get(ConfirmService);
+    const networkMemberService: NetworkMemberService = TestBed.get(NetworkMemberService);
```
redondant types?

cexbrayat commented on frontend/src/app/person-network-members/person-network-members.component.spec.ts:
```diff
@@ -0,0 +1,257 @@
+
+  it('should not delete a member and reload list if not confirmed', () => {
+    const confirmService: ConfirmService = TestBed.get(ConfirmService);
+    const networkMemberService: NetworkMemberService = TestBed.get(NetworkMemberService);
```
same

cexbrayat commented on frontend/src/app/person-network-members/person-network-members.component.spec.ts:
```diff
@@ -0,0 +1,257 @@
+    expect(textArea.value).toBe('');
+  });
+
+  it('should hide the form and re-display the list chen cancelling', () => {
```
chen -> when

cexbrayat commented on frontend/src/app/person-network-members/person-network-members.component.spec.ts:
```diff
@@ -0,0 +1,257 @@
+    createButton.click();
+    fixture.detectChanges();
+
+    const networkMemberService: NetworkMemberService = TestBed.get(NetworkMemberService);
```
same

cexbrayat commented on frontend/src/app/person-network-members/person-network-members.component.spec.ts:
```diff
@@ -0,0 +1,257 @@
+  });
+
+  it('should create and reload the list', () => {
+    const networkMemberService: NetworkMemberService = TestBed.get(NetworkMemberService);
```
same

cexbrayat commented on frontend/src/app/person-network-members/person-network-members.component.spec.ts:
```diff
@@ -0,0 +1,257 @@
+    const textArea = element.querySelector('#text') as HTMLTextAreaElement;
+    expect(textArea.value).toBe('Dr. No');
+
+    const networkMemberService: NetworkMemberService = TestBed.get(NetworkMemberService);
```
same

cexbrayat commented on frontend/src/app/person-network-members/person-network-members.component.ts:
```diff
@@ -0,0 +1,74 @@
+import { NETWORK_MEMBER_TYPE_TRANSLATIONS } from '../display-network-member-type.pipe';
+import { FormBuilder, FormGroup, Validators } from '@angular/forms';
+import { NetworkMemberCommand } from '../models/network-member.command';
+import { Observable } from 'rxjs';
```
nit: I personally don't like to have imports mixed up. The recommandation is to have angular imports, followed by 3rd party imports, followed by a blank line, and then the imports from our codebase.

jnizet replied:
I there some IntelliJ option to do that? I never write imports or arrange them. I just let the IDE add them and optimize them.

cexbrayat replied:
Sadly, I think there is not (AFAIK). Maybe we could find a TSLint rule for that. I usually reorder by hand when I'm done...

cexbrayat commented on frontend/src/app/person-network-members/person-network-members.component.ts:
```diff
@@ -0,0 +1,74 @@
+    }).pipe(
+      switchMap(() => this.networkMemberService.delete(this.person.id, member.id)),
+      switchMap(() => this.networkMemberService.list(this.person.id)),
+    ).subscribe(members => this.members = members, () => {});
```
Is the empty error handler necessary?

jnizet replied:
Yes. I chose to emit an error when the use doesn't confirm. If I don't catch that error, the test testing what happens when the user doesn't confirm blows up, and the error also appears in the console, which is ugly.

I could have chosen to emit true/false, but then I'd have to remember to filter true values every time I use confirm, which is worse, IMO.

jnizet commented on frontend/src/app/network-members-resolver.service.spec.ts:
```diff
@@ -0,0 +1,35 @@
+
+  it('should resolve network members', () => {
+    const resolver: NetworkMembersResolverService = TestBed.get(NetworkMembersResolverService);
+    const service: NetworkMemberService = TestBed.get(NetworkMemberService);
```
Unfortunately, no. I *think* it used to return a value of the same type as the first argument, but it now returns any (I guess to deal with tokens that are not types). So I could leave out the type, but then I'd have any, so no code completion, etc.

cexbrayat replied:
Damn, I did not think about that. That means I have to add them in our ebook and training too...

cexbrayat replied:
After digging into Angular source code history, I'm starting to think that I hallucinated the fact that `TestBed.get(T)` ever returned `T`...

jnizet replied:
That said, it would be a nice amendment to the API, because having to repeat the type every time is cumbersome. Something like
```
get<T>(typeToken: { new(...args: any[]): T }): T {
  return TestBed.get(typeToken);
}
```
or, if they don't want an overload, 
```
getByType<T>(type: { new(...args: any[]): T }): T {
  return TestBed.get(typeToken);
}
```

We could of course define such a global function ourselves, but it would feel more natural if it was in TestBed.

cexbrayat replied:
I've been trying to come up with a PR since my comment :) I was trying to "fix" the signature (because `Injector.get(T)` *does* return `T`) but it doesn't work. I'll try an overload.

acrepet commented:
I've just tested :+1: 
except this error : If you forget to fill the type (lwyer, doctor) of member : we have a validation error message  (type is required), but if you have an existing member and want to update it, and if you don't choose a type : we do not have a message error but an exception: 

"Une erreur technique inattendue s'est produite. Essayez de recharger la page. 
400 - JSON parse error: Cannot deserialize value of type `org.globe42.domain.NetworkMemberType` from String "null": value not one of declared Enum instance names: [OTHER, LAWYER, SOCIAL_ASSISTANT, HELPER, DOCTOR, PERSON_TO_WARN]; nested exception is com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `org.globe42.domain.NetworkMemberType` from String "null": value not one of declared Enum instance names: [OTHER, LAWYER, SOCIAL_ASSISTANT, HELPER, DOCTOR, PERSON_TO_WARN] at [Source: (PushbackInputStream); line: 1, column: 9] (through reference chain: org.globe42.web.persons.NetworkMemberCommandDTO["type"])"

acrepet commented on frontend/src/app/person-network-members/person-network-members.component.html:
```diff
@@ -0,0 +1,50 @@
+  <div class="form-group">
+    <label for="type">Type</label>
+    <select class="form-control" id="type" formControlName="type">
+      <option [value]="null"></option>
```
Why a null option if the type is mandatory?

jnizet commented:
@acrepet good catch. I fixed it. The blank option is mainly useful for the creation case, where I want the user to consciously choose a type, rather than blindly seclecting the first possible type. I could remove it in edition mode, but I think it doesn't hurt much to leave it there.

acrepet commented:
OK, I've tested :+1: 
(just a stupid question : I think that you've shashed your commits, so how I can review your last commit which fix the error?)

jnizet commented:
The question is not stupid at all. I'll try to only squash before merging to master next time. I changed

`<option [value]="null"></option>`

to

`<option></option>`.

The test that checked that the selected index was 0 by default now checks that it's -1.

And in the last test, which checks the update and the save, I added a few lines setting the selected index to 0, checking that the error message appears when submitting, then change again and check it disappears.

acrepet commented:
OK thanks @jb I've got it!
